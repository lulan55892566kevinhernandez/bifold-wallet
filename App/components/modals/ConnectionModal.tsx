import React, { useContext, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Dimensions, Modal, StyleSheet, Text, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'

import ConnectionPending from '../../assets/img/connection-pending.svg'
import { Context } from '../../store/Store'
import { ColorPallet, Colors, TextTheme } from '../../theme'
import Button, { ButtonType } from '../buttons/Button'

const { height } = Dimensions.get('window')
const connectionTimerDelay = 3000 // in ms
const imageDisplayOptions = {
  fill: Colors.text,
  height: 250,
  width: 250,
}

const styles = StyleSheet.create({
  container: {
    minHeight: height,
    flexDirection: 'column',
    paddingHorizontal: 25,
    backgroundColor: ColorPallet.brand.primaryBackground,
  },
  image: {
    marginTop: 40,
  },
  messageContainer: {
    alignItems: 'center',
    marginTop: 54,
  },
  messageText: {
    fontWeight: 'normal',
    textAlign: 'center',
  },
  delayMessageContainer: {
    marginTop: 30,
  },
  delayMessageText: {
    textAlign: 'center',
    marginBottom: 30,
  },
})

const ConnectionModal: React.FC = () => {
  const { t } = useTranslation()
  const [modalVisible, setModalVisible] = useState<boolean>(false)
  const [shouldShowDelayMessage, setShouldShowDelayMessage] = useState<boolean>(false)
  const [state] = useContext(Context)
  let timer: NodeJS.Timeout

  const onDismissModalTouched = () => {
    setShouldShowDelayMessage(false)
    setModalVisible(false)
  }

  useEffect(() => {
    setModalVisible(state.notifications.ConnectionPending)

    if (state.notifications.ConnectionPending) {
      timer = setTimeout(() => {
        setShouldShowDelayMessage(true)
      }, connectionTimerDelay)
    }

    return () => {
      timer && clearTimeout(timer)
    }
  }, [state])

  return (
    <Modal visible={modalVisible} transparent={true}>
      <SafeAreaView style={[styles.container]}>
        <View style={[styles.messageContainer]}>
          <Text style={[TextTheme.headingThree, styles.messageText]}>{t('Connection.JustAMoment')}</Text>
          <ConnectionPending style={[styles.image]} {...imageDisplayOptions} />
          {shouldShowDelayMessage && (
            <View style={[styles.delayMessageContainer]}>
              <Text style={[TextTheme.normal, styles.delayMessageText]}>{t('Connection.TakingTooLong')}</Text>
              <Button
                title={t('Connection.BackToHome')}
                accessibilityLabel={t('Global.Home')}
                onPress={onDismissModalTouched}
                buttonType={ButtonType.Secondary}
              />
            </View>
          )}
        </View>
      </SafeAreaView>
    </Modal>
  )
}

export default ConnectionModal
