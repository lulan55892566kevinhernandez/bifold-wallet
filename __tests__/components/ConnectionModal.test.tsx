import { render, waitFor, fireEvent } from '@testing-library/react-native'
import React, { useContext } from 'react'

import ConnectionModal from '../../App/components/modals/ConnectionModal'

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}))

jest.useFakeTimers('legacy')
jest.spyOn(global, 'setTimeout')

const setState = jest.fn()
const state = {
  onboarding: {
    DidAgreeToTerms: false,
    DidCompleteTutorial: false,
    DidCreatePIN: false,
  },
  notifications: {
    ConnectionPending: true,
  },
  error: null,
}

// @ts-ignore
useContext.mockImplementation(() => [state, setState])

describe('ConnectionModal Component', () => {
  test('Renders correctly', async () => {
    const tree = render(<ConnectionModal />)

    expect(tree).toMatchSnapshot()
    expect(tree.queryByText('Connection.BackToHome')).toBeNull()
  })

  test('Updates after delay', async () => {
    const tree = render(<ConnectionModal />)

    await waitFor(() => {
      jest.runAllTimers()
    })

    expect(tree).toMatchSnapshot()
  })

  test('Dismiss on demand', async () => {
    const tree = render(<ConnectionModal />)

    await waitFor(() => {
      jest.runAllTimers()
    })

    const backHomeBtn = await tree.findByText('Connection.BackToHome')
    fireEvent(backHomeBtn, 'press')

    expect(tree).toMatchSnapshot()
  })
})
