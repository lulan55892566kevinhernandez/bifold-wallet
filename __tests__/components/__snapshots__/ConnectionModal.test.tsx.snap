// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionModal Component Dismiss on demand 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={false}
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flexDirection": "column",
          "minHeight": 1334,
          "paddingHorizontal": 25,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginTop": 54,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 26,
              "fontWeight": "bold",
            },
            Object {
              "fontWeight": "normal",
              "textAlign": "center",
            },
          ]
        }
      >
        Connection.JustAMoment
      </Text>
      <
        fill="#FFFFFF"
        height={250}
        style={
          Array [
            Object {
              "marginTop": 40,
            },
          ]
        }
        width={250}
      />
    </View>
  </RNCSafeAreaView>
</Modal>
`;

exports[`ConnectionModal Component Renders correctly 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flexDirection": "column",
          "minHeight": 1334,
          "paddingHorizontal": 25,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginTop": 54,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 26,
              "fontWeight": "bold",
            },
            Object {
              "fontWeight": "normal",
              "textAlign": "center",
            },
          ]
        }
      >
        Connection.JustAMoment
      </Text>
      <
        fill="#FFFFFF"
        height={250}
        style={
          Array [
            Object {
              "marginTop": 40,
            },
          ]
        }
        width={250}
      />
    </View>
  </RNCSafeAreaView>
</Modal>
`;

exports[`ConnectionModal Component Updates after delay 1`] = `
<Modal
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "backgroundColor": "#000000",
          "flexDirection": "column",
          "minHeight": 1334,
          "paddingHorizontal": 25,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "marginTop": 54,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 26,
              "fontWeight": "bold",
            },
            Object {
              "fontWeight": "normal",
              "textAlign": "center",
            },
          ]
        }
      >
        Connection.JustAMoment
      </Text>
      <
        fill="#FFFFFF"
        height={250}
        style={
          Array [
            Object {
              "marginTop": 40,
            },
          ]
        }
        width={250}
      />
      <View
        style={
          Array [
            Object {
              "marginTop": 30,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 18,
                "fontWeight": "normal",
              },
              Object {
                "marginBottom": 30,
                "textAlign": "center",
              },
            ]
          }
        >
          Connection.TakingTooLong
        </Text>
        <View
          accessibilityLabel="Global.Home"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "borderColor": "#35823F",
              "borderRadius": 4,
              "borderWidth": 2,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#35823F",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "textAlign": "center",
                },
                false,
              ]
            }
          >
            Connection.BackToHome
          </Text>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</Modal>
`;
